### ISSUE TITLE
#2971 performance reduction for large dataframes

### ISSUE THREAD

**Issue by andrewhill157** [marimo-team/marimo#3107]
### Describe the bug

As far as I can tell the commit merging #2971 resulted in a significant performance reduction for basic dataframe operations at least for large dataframes (lots of columns)? I'm using polars here.

For the example notebook below if I run on 700e0aa697cc6a6ba698946c417bd529d6f4b0c7 the final two cells run much slower than on the previous commit 4aeb8eecda74531f992a09eeac4f97c933785528. On my machine they are more or less instant on 4aeb8eecda74531f992a09eeac4f97c933785528 and take well over a second on 700e0aa697cc6a6ba698946c417bd529d6f4b0c7.

I think this behavior is consistent on my end, but let me know if you have any trouble reproducing! Thanks for the help!

### Environment

```
{
  "marimo": "0.9.25",
  "OS": "Darwin",
  "OS Version": "24.0.0",
  "Processor": "arm",
  "Python Version": "3.10.4",
  "Binaries": {
    "Browser": "--",
    "Node": "v22.9.0"
  },
  "Dependencies": {
    "click": "8.1.3",
    "docutils": "0.21.2",
    "itsdangerous": "2.2.0",
    "jedi": "0.19.1",
    "markdown": "3.5.2",
    "narwhals": "1.16.0",
    "packaging": "23.2",
    "psutil": "5.9.8",
    "pygments": "2.17.2",
    "pymdown-extensions": "10.7",
    "pyyaml": "6.0.1",
    "ruff": "0.6.7",
    "starlette": "0.37.2",
    "tomlkit": "0.12.3",
    "typing-extensions": "4.12.2",
    "uvicorn": "0.29.0",
    "websockets": "12.0"
  },
  "Optional Dependencies": {
    "altair": "5.4.1",
    "duckdb": "1.1.3",
    "pandas": "2.2.2",
    "polars": "1.7.1",
    "pyarrow": "17.0.0"
  }
}
```

### Code to reproduce

```
import marimo

__generated_with = "0.9.25"
app = marimo.App(width="medium")


@app.cell
def __():
    import polars as pl
    import numpy as np
    from typing import Tuple
    import marimo as mo

    def create_test_dataframe(num_rows: int) -> Tuple[pl.DataFrame, list[str]]:
        """
        Create a large test dataframe with various column types including 1200 float columns.

        Args:
            num_rows: Number of rows to generate

        Returns:
            Tuple of (DataFrame, list of group values)
        """
        # Set random seed for reproducibility
        np.random.seed(42)

        # Generate random data for different column types
        data = {
            "int_col1": np.random.randint(0, 1000000, num_rows),
            "int_col2": np.random.randint(-1000000, 1000000, num_rows),
            "float_col1": np.random.normal(0, 1, num_rows),
            "float_col2": np.random.uniform(0, 100, num_rows),
            "str_col1": np.random.choice(['A', 'B', 'C', 'D', 'E'], num_rows),
            "str_col2": [f"val_{i}" for i in np.random.randint(0, 1000, num_rows)],
            "bool_col1": np.random.choice([True, False], num_rows),
            "bool_col2": np.random.choice([True, False], num_rows),
            "group": np.random.choice(['group1', 'group2', 'group3', 'group4'], num_rows),
        }

        # Add 1200 float columns with different distributions
        for i in range(1200):
            if i % 3 == 0:
                # Normal distribution
                data[f"float_normal_{i}"] = np.random.normal(0, 1, num_rows)
            elif i % 3 == 1:
                # Uniform distribution
                data[f"float_uniform_{i}"] = np.random.uniform(-100, 100, num_rows)
            else:
                # Exponential distribution
                data[f"float_exp_{i}"] = np.random.exponential(1, num_rows)

        # Create the original derived columns
        data.update({
            f"derived_col_{i}": data["int_col1"] + np.random.randint(0, 100, num_rows)
            for i in range(10)
        })

        df = pl.DataFrame(data)
        unique_groups = df["group"].unique().to_list()

        return df
    return Tuple, create_test_dataframe, mo, np, pl


@app.cell
def __(create_test_dataframe):
    num_rows = 1000
    df = create_test_dataframe(num_rows)
    return df, num_rows


@app.cell
def __(df, pl):
    x = df.filter(pl.col("group") == "group2")
    return (x,)


if __name__ == "__main__":
    app.run()

```

**Comment by mscolnick** [comment]
Thanks for the repro. I will investigate right now

**Comment by mscolnick** [comment]
Thanks, i can repro with:
`uvx --with="marimo==0.9.26" marimo edit 3107-dataframe-perf.py` # slow
and 
`uvx --with="marimo==0.9.25" marimo edit 3107-dataframe-perf.py` # fast

**Comment by mscolnick** [comment]
I found the issue (great catch by the way!) and thank you for identifying the offending PR. 

I am going to find a few other places this could arise (in pandas, etc), merge, cut a release. and then i will followup with some performance tests to avoid a regression

**Comment by andrewhill157** [comment]
Awesome, thank you!

**Comment by mscolnick** [comment]
@andrewhill157 i released `0.9.34`

**Comment by andrewhill157** [comment]
thanks for the super quick fix!
